### PR TITLE
Update Detection

### PR DESCRIPTION
**Pull request by Animatory**

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -14,17 +14,18 @@ dependencies:
   - pytorch=1.12.1
   - torchvision=0.13.1
   - cudatoolkit=11.6
-  - pytorch-lightning=1.7.7
-  - torchmetrics=0.9.2
   - parameterized=0.8.1
   - pip:
+      - pytorch-lightning==1.8.4.post0
```
Comment (VladVin) [inline environment.yml:19]: Why is the version with postfix?

File: examples/configs/classification_cifar10.yaml [hunk 2]
```diff
@@ -43,7 +43,7 @@ data:
           input_dtype: *input_dtype
           train: true
           download: true
-          data_folder: &data_folder ${oc.env:HOME}/.cache/torchok/data/cifar10
+          data_folder: &data_folder ${oc.env:HOME}/cifar10/data
```
Comment (VladVin) [inline examples/configs/classification_cifar10.yaml:46]: It's not good to create many folders in user's home directory, one folder is much more convenient for a user

File: examples/configs/classification_cifar10.yaml [hunk 3]
```diff
@@ -89,8 +89,8 @@ seed_params:
   workers: true
 
 logger:
-  log_dir: '${oc.env:HOME}/.cache/torchok/logs/'
-  experiment_name: cifar10
+  log_dir: '${oc.env:HOME}/cifar10/logs'
+  experiment_name: resnet18
```
Comment (VladVin) [inline examples/configs/classification_cifar10.yaml:93]: It's not just `resnet18`, it's `cifar10_resnet18`

File: examples/configs/classification_cifar10_mlflowx.yaml [hunk 4]
```diff
@@ -94,15 +94,19 @@ seed_params:
...
-  log_dir: '${oc.env:HOME}/.cache/torchok/logs/'
   experiment_name: cifar10
+  log_dir: '${oc.env:HOME}/${logger.experiment_name}/logs'
   name: MLFlowLoggerX
+  timestamp: '${now:%Y-%m-%d}/${now:%H-%M-%S}'
   params:
-    artifact_location: 'file://${oc.env:HOME}/.cache/torchok/logs/artifacts/'
+    run_name: resnet18
+    tracking_uri: 'mysql+pymysql://${oc.env:DB_USER}:${oc.env:DB_PASSWORD}@${oc.env:DB_HOST}/mlflow'
+    save_dir: '${logger.log_dir}/${logger.params.run_name}/${logger.timestamp}'
```
Comment (VladVin) [inline examples/configs/classification_cifar10_mlflowx.yaml:104]: Why `save_dir` and `log_dir` are different?

File: torchok/models/necks/detection/fpn.py [hunk 5]
```diff
@@ -60,3 +61,57 @@ class FPN(necks.FPN):
     def __init__(self, in_channels, **kwargs):
         in_channels = list(in_channels[::-1])
         super(FPN, self).__init__(in_channels=in_channels, **kwargs)
+
+    def forward(self, inputs):
```
Comment (VladVin) [inline torchok/models/necks/detection/fpn.py:65]: Please, move commits related to detection to a separate PR

**Review by VladVin — CHANGES_REQUESTED** [review]

File: environment.yml [hunk 6]
```diff
@@ -5,7 +5,7 @@ channels:
   - defaults
 dependencies:
```
Comment (VladVin) [inline environment.yml:6]: Where are updates to `pyproject.toml`?

File: environment.yml (hunk 1 above)
Comment (Animatory) [reply environment.yml:19]: [Check here](https://github.com/Lightning-AI/lightning/releases)

**Review by VladVin — CHANGES_REQUESTED** [review]

File: examples/configs/classification_cifar10.yaml (hunk 2 above)
Comment (Animatory) [reply examples/configs/classification_cifar10.yaml:46]: As a user, it is not convinient what you suggest. As I can set up as a `oc.env:HOME` anything I want, therefore I can choose where all logs will be

File: examples/configs/classification_cifar10.yaml (hunk 3 above)
Comment (Animatory) [reply examples/configs/classification_cifar10.yaml:93]: No, cifar10 is located at higher hierarchy, therefore there is no need to duplicate information about cifar10

File: examples/configs/classification_cifar10_mlflowx.yaml (hunk 4 above)
Comment (Animatory) [reply examples/configs/classification_cifar10_mlflowx.yaml]: Remove save_dir as it is not used

File: examples/configs/classification_cifar10.yaml (hunk 2 above)
Comment (VladVin) [reply examples/configs/classification_cifar10.yaml:46]: Usually, `$HOME` env variable is set by the system since it's a common practice to point to the home directory. A user won't change this variable to have better experience with our library because it introduces a mess...

File: examples/configs/classification_cifar10.yaml (hunk 3 above)
Comment (VladVin) [reply examples/configs/classification_cifar10.yaml:93]: I don't see an experiment `cifar10`, what do you mean?

File: examples/configs/classification_cifar10_mlflowx.yaml (hunk 4 above)
Comment (VladVin) [reply examples/configs/classification_cifar10_mlflowx.yaml]: I don't see changes

**Review by VladVin — CHANGES_REQUESTED** [review]

File: torchok/constructor/logger.py [hunk 7]
```diff
@@ -103,23 +131,20 @@ def __init__(self,
...
-        status = 'FINISHED' if status == 'success' else status
-
         upload_file_paths = chain(self._save_dir.glob('*.ckpt'), self._save_dir.glob('*.onnx'))
         for file_path in upload_file_paths:
-            self.experiment.log_artifact(self._run_id, file_path.as_posix())
+            self.experiment.log_artifact(self.run_id, file_path.as_posix())
 
-        if self.experiment.get_run(self._run_id):
-            self.experiment.set_terminated(self._run_id, status)
+        super().finalize(status)
```
Comment (PososikTeam) [inline torchok/constructor/logger.py:147]: Is status write correct now?

File: torchok/constructor/logger.py [hunk 8]
```diff
@@ -1,19 +1,46 @@
...
 from pytorch_lightning.loggers.neptune import NeptuneLogger
 from pytorch_lightning.loggers.tensorboard import TensorBoardLogger
 from pytorch_lightning.loggers.wandb import WandbLogger
+from pytorch_lightning.utilities import rank_zero_warn
 from pytorch_lightning.utilities.logger import _convert_params
 
-from argparse import Namespace
-from omegaconf.listconfig import ListConfig
+
+def create_logger(logger_config) -> Logger:
```
Comment (PososikTeam) [inline torchok/constructor/logger.py:19]: can you please write documentation for this function

File: torchok/data/datasets/detection/detection.py [hunk 9]
```diff
@@ -141,7 +142,7 @@ def get_raw(self, idx: int) -> dict:
         return sample
 
     def filter_bboxes(self, bboxes, labels, shape):
```
Comment (PososikTeam) [inline torchok/data/datasets/detection/detection.py:144]: i think need documentation for each function

File: torchok/data/datasets/detection/detection.py [hunk 10]
```diff
@@ -169,11 +170,11 @@ def __getitem__(self, idx: int) -> dict:
 
     def collate_fn(self, batch):
         r"""Puts each data field into a tensor with outer dimension batch size"""
```
Comment (PososikTeam) [inline torchok/data/datasets/detection/detection.py:172]: I think you need to add torch issue in documentation to explain why we need tensors

File: torchok/models/backbones/mmdet_backbones/resnet.py [hunk 11]
```diff
@@ -658,6 +641,10 @@ def _freeze_stages(self):
             for param in m.parameters():
                 param.requires_grad = False
 
+        for m in self.modules():
```
[the remark this answers is not in the document]
Comment (PososikTeam) [reply torchok/models/backbones/mmdet_backbones/resnet.py:644]: Also what you think about rename `self.norm_eval` to like in FreezeUnfreeze Callback `self.bn_track_running_stats`. Also why we need this if we have FreezeUnfreeze Callback?

File: torchok/models/necks/detection/fpn.py (hunk 5 above)
Comment (PososikTeam) [inline torchok/models/necks/detection/fpn.py:65]: Where the diffrence with https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/necks/fpn.py ?

File: torchok/tasks/detection.py [hunk 12]
```diff
@@ -68,15 +74,18 @@ def forward_with_gt(self, batch: Dict[str, Union[torch.Tensor, int]]) -> Dict[st
...
 
-        if 'bboxes' in batch:
-            output['gt_bboxes'] = batch.get('bboxes')
-            output['gt_labels'] = batch.get('label')
+        if 'bboxes' in batch.keys():
+            if 'num_bboxes' in batch.keys():
+                output['gt_bboxes'] = [bbox[:n, :4] for bbox, n in zip(batch['bboxes'], batch['num_bboxes'])]
+                output['gt_labels'] = [bbox[:n, 4].long() for bbox, n in zip(batch['bboxes'], batch['num_bboxes'])]
+            else:
+                output['gt_bboxes'] = batch.get('bboxes')[:, :, :4]
```
Comment (PososikTeam) [inline torchok/tasks/detection.py:87]: Is it COCO dataset has't `num_bboxes`? If yes why it has't?

**Review by PososikTeam — COMMENTED** [review]
Please answer to my questions

File: torchok/models/backbones/mmdet_backbones/resnet.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/models/backbones/mmdet_backbones/resnet.py:644]: This is from mmdet just to check if the model is correct. Now when we reproduced results with our own backbone we can remove mmresnet. What do you think?

File: torchok/models/necks/detection/fpn.py (hunk 5 above)
Comment (Animatory) [reply torchok/models/necks/detection/fpn.py:65]: Here was auto push to fp16 which I don't want to be always enabled

File: torchok/tasks/detection.py (hunk 12 above)
Comment (Animatory) [reply torchok/tasks/detection.py:87]: There may be another dataset

File: torchok/models/backbones/mmdet_backbones/resnet.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/models/backbones/mmdet_backbones/resnet.py]: Removed it

File: torchok/constructor/logger.py (hunk 8 above)
Comment (Animatory) [reply torchok/constructor/logger.py:19]: Added

**Review by PososikTeam — APPROVED** [review]

**Review by PososikTeam — DISMISSED** [review]

**Review by VladVin — CHANGES_REQUESTED** [review]

File: torchok/tasks/detection.py [hunk 13]
```diff
@@ -68,8 +68,7 @@ def forward_with_gt(self, batch: Dict[str, Union[torch.Tensor, int]]) -> Dict[st
         """Forward with ground truth labels."""
```
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/tasks/detection.py:68]: It is used inside `self.bbox_head.get_bboxes`. I updated forward function

File: torchok/tasks/base.py [hunk 14]
```diff
@@ -117,6 +118,11 @@ def on_test_start(self) -> None:
         if self._hparams.task.load_checkpoint is not None:
             load_checkpoint(self, **self._hparams.task.load_checkpoint)
 
+    def transfer_batch_to_device(self, batch: Any, device: torch.device, dataloader_idx: int) -> Any:
+        if isinstance(batch, pd.Series):
```
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/tasks/base.py:122]: I need this to test output in form of pandas series that is equal to dict

File: torchok/data/datasets/examples/coco_detection.py [hunk 15]
```diff
@@ -191,3 +192,24 @@ def _download(self) -> None:
...
+
+    def __getitem__(self, idx: int) -> dict:
+        """Get item sample.
+
+        Returns:
+            sample: dict, where
+            sample['image'] - Tensor, representing image after augmentations and transformations, dtype=input_dtype.
+            sample['target'] - Target class or labels, dtype=target_dtype.
+            sample['bboxes'] - Target bboxes, dtype=bbox_dtype.
+            sample['index'] - Index.
```
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/data/datasets/examples/coco_detection.py:204]: It is the same as input idx

File: torchok/data/datasets/detection/detection.py [hunk 16]
```diff
@@ -132,19 +134,30 @@ def get_raw(self, idx: int) -> dict:
...
+        """Filter empty bounding boxes.
+
+        Args:
+            bboxes: List of bounding box.
+            labels: array of bbox labels
+            rows: Image height.
+            cols: Image width.
+
+        Returns:
+            List of bounding boxes.
```
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/data/datasets/detection/detection.py:155]: Fixed

File: torchok/constructor/logger.py [hunk 17]
```diff
@@ -1,19 +1,50 @@
...
+        experiment_subdir = str(full_outputs_path.relative_to(experiment_path))
+
+        logger = build_logger(logger_class_name=logger_config.name,
+                              logger_class_params=logger_config.params,
+                              outputs_path=logger_config.log_dir,
+                              experiment_name=logger_config.experiment_name,
+                              experiment_subdir=experiment_subdir,
+                              full_outputs_path=full_outputs_path)
+
+        if os.environ.get('LOCAL_RANK') is not None:
```
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/constructor/logger.py:44]: It is used for ddp. Added comment

File: examples/configs/detection/fcos_resnet50_coco.yaml [hunk 18]
```diff
@@ -35,8 +35,9 @@ task:
           iou_threshold: 0.5
```
[the remark this answers is not in the document]
Comment (Animatory) [reply examples/configs/detection/fcos_resnet50_coco.yaml:35]: No, I want to keep various configs for different tasks. Now there is only one, but with addition of new architectures there will be more examples

File: examples/configs/classification_cifar10.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Animatory) [reply examples/configs/classification_cifar10.yaml:46]: Renamed it from HOME to DATA_DIR. Now it will not confuse anyone

File: examples/configs/onnx_infer.yaml [hunk 19]
```diff
@@ -18,7 +18,7 @@ data:
         params:
           train: false
           download: true
-          data_folder: &data_folder ${oc.env:HOME}/.cache/torchok/data/cifar10
+          data_folder: &data_folder ${oc.env:DATA_DIR}/.cache/torchok/data/cifar10
```
Comment (VladVin) [inline examples/configs/onnx_infer.yaml:21]: Who is setting `DATA_DIR`? Where is the instruction?!

File: torchok/constructor/logger.py [hunk 20]
```diff
@@ -1,19 +1,50 @@
...
+
+def create_logger(logger_config: DictConfig) -> Logger:
+    """
+    Create logger based on logger config.
+    """
+    if logger_config is not None:
+        if logger_config.name == 'MLFlowLoggerX':
+            run_name = logger_config.params.run_name
+        else:
+            run_name = logger_config.experiment_name
```
[the remark this answers is not in the document]
Comment (VladVin) [reply torchok/constructor/logger.py:28]: Don't call it `run_name` then. Let's have `experimental_name` and an optional `run_name`. Change your paths processing so that the full path is constructed as `log_dir/experimental_name` or `log_dir/experimental_name/run_name` depending on either `run_name` is specified. The current logic is unclear, let's make everything transparent

File: torchok/constructor/logger.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (VladVin) [reply torchok/constructor/logger.py:45]: Describe also, what `LOCAL_RANK = None` means

File: torchok/data/datasets/examples/coco_detection.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (VladVin) [reply torchok/data/datasets/examples/coco_detection.py:204]: Why do you return it? I asked to explain, don't skip my requests please

File: torchok/tasks/base.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (VladVin) [reply torchok/tasks/base.py:122]: We don't support pandas objects as outputs from datasets. Don't bring the mess please

File: torchok/tasks/detection.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (VladVin) [reply torchok/tasks/detection.py:69]: But how did you calculate the validation metrics if you skipped NMS initially?

**Review by VladVin — CHANGES_REQUESTED** [review]

File: torchok/tasks/detection.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Animatory) [reply torchok/tasks/detection.py:69]: Check `training_step`, `validation_step` and `test_step`. You will see that `self.bbox_head.get_bboxes` called inside these functions

**Review by VladVin — APPROVED** [review]
